### PR TITLE
Delete __if_exists

### PR DESCRIPTION
**Pull request by am11**
https://devblogs.microsoft.com/oldnewthing/20190828-00/?p=102812

### CODE REVIEW

File: src/coreclr/inc/check.h [hunk 1]
```diff
@@ -16,6 +16,14 @@
 #include "daccess.h"
 #include "unreachable.h"
 
+// Macros for creating type traits to check if a member exists
+#define DEFINE_MEMBER_EXISTENCE_CHECK(Member) \
+template<typename T, typename = void> \
+struct has_##Member : std::false_type {}; \
+\
+template<typename T> \
+struct has_##Member<T, decltype((void)&T::Member, void())> : std::true_type {};
```
Comment (jkoritzinsky) [inline src/coreclr/inc/check.h:25]: Can we use the detection idiom here instead? (We have a usage of it already in dacenumerablehash.inl, we can move some of that here.)

File: src/coreclr/inc/check.h [hunk 2]
```diff
@@ -282,23 +290,36 @@ do                                                                  \
...
+    CHECK(obj.Invariant());
+}
+
+template <typename TYPENAME>
+CHECK CheckInvariiantOnly(TYPENAME &obj, std::false_type) { CHECK_OK; }
+
+template <typename TYPENAME>
+CHECK CheckInternalInvariiantOnly(TYPENAME &obj, std::true_type)
+{
+    CHECK(obj.InternalInvariant());
```
Comment (am11) [inline src/coreclr/inc/check.h:317]:
```suggestion
    CHECK(obj.InternalInvariant());
    CHECK_OK;
```

File: src/coreclr/inc/check.h [hunk 3]
```diff
@@ -282,23 +290,36 @@ do                                                                  \
...
-    }
-#endif
+    CheckInvariiantOnly(obj, has_Invariant<TYPENAME>());
+    CheckInternalInvariiantOnly(obj, has_InternalInvariant<TYPENAME>());
 
     CHECK_OK;
 }
 
+template <typename TYPENAME>
+CHECK CheckInvariiantOnly(TYPENAME &obj, std::true_type)
```
Comment (mangod9) [inline src/coreclr/inc/check.h:306]: nit: typo?

File: src/coreclr/inc/check.h (hunk 1 above)
Comment (am11) [reply src/coreclr/inc/check.h:25]: Could you point me to it? Is it: https://github.com/dotnet/runtime/blob/f0edcd8ac87ff9f7f71d5cc3faf7429c18734b55/src/coreclr/vm/dacenumerablehash.inl#L475

File: src/coreclr/inc/check.h [hunk 4]
```diff
@@ -282,23 +290,37 @@ do                                                                  \
...
+    CheckInvariantOnly(obj, has_Invariant<TYPENAME>());
+    CheckInternalInvariantOnly(obj, has_InternalInvariant<TYPENAME>());
+
+    CHECK_OK;
+}
+
+template <typename TYPENAME>
+CHECK CheckInvariantOnly(TYPENAME &obj, std::true_type)
+{
+    CHECK(obj.Invariant());
```
Comment (am11) [inline src/coreclr/inc/check.h:308]:
```suggestion
    CHECK(obj.Invariant());
    CHECK_OK;
```

File: src/coreclr/inc/check.h (hunk 1 above)
Comment (am11) [reply src/coreclr/inc/check.h:25]: I implemented the C++17 version first, and it was quite straightforward:

```
#define EXECUTE_IF_MEMBER_EXISTS(Type, Member, CodeBlock) \
if constexpr (has_##Member<Type>::value) { \
    CodeBlock \
}
```

but then this enable_if(cond), enable_if(!cond) clutter came to support std=c++11. We can definitely simplify some of this code (revert parts of https://github.com/dotnet/runtime/commit/856e8c3510cb0aaaa3b1a1679552c03ae5308b9e) when we are ready to drop support for older compilers. 😅

---

I think the some parts of the product had issues even with clang-9 when we had to switch to clang-12+. Not sure if build with old compiler make sense given those run-time issues; especially when we support the old libc with latest toolchain via cross-compilation? Perhaps it's already the case that https://github.com/dotnet/runtime/blob/1ee61b8667d9db31c0fe53d2e1abc274fb486cfc/eng/common/native/init-compiler.sh#L66-L67
should be:
 ```sh
if [ "$compiler" = "clang" ]; then versions="18 17 16 15 14 13" 
elif [ "$compiler" = "gcc" ]; then versions="14 13 12"; fi
```
for more than one reason and simplify clang<v13 checks (https://github.com/dotnet/runtime/pull/102278/files) even further.

File: src/coreclr/inc/check.h [hunk 5]
```diff
@@ -16,6 +16,19 @@
 #include "daccess.h"
 #include "unreachable.h"
 
+// Use the C++ detection idiom (https://isocpp.org/blog/2017/09/detection-idiom-a-stopgap-for-concepts-simon-brand) to call the
+// derived table's EnumMemoryRegionsForEntry method if it defines one.
```
Comment (jkoritzinsky) [inline src/coreclr/inc/check.h:20]:
```suggestion
// Use the C++ detection idiom (https://isocpp.org/blog/2017/09/detection-idiom-a-stopgap-for-concepts-simon-brand)
```

File: src/coreclr/inc/iterator.h [hunk 6]
```diff
@@ -238,30 +239,34 @@ class CheckedIteratorBase
...
 
-    void Resync(const CONTAINER *container)
+    void Init(const CONTAINER *container, std::false_type) { /* No-op */ }
+
+    template<typename Dummy = CONTAINER>
+    typename std::enable_if<has_m_revision<Dummy>::value, void>::type Resync(const CONTAINER *container)
     {
         LIMITED_METHOD_DAC_CONTRACT;
 
 #if defined(_DEBUG) && (defined(_MSC_VER) || defined(__llvm__))
```
Comment (am11) [inline src/coreclr/inc/iterator.h:262]:
```suggestion
#if defined(_DEBUG)
```

**Review by jkoritzinsky — APPROVED** [review]

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by am11** [comment]
@jkoritzinsky, could this be merged? Failures look unrelated.

**Pull request merged**